### PR TITLE
[stable/traefik] Basic auth support for dashboard

### PR DESCRIPTION
**Pull request by cknowles**

### CODE REVIEW

**Comment by viglesiasce** [comment]
@k8s-bot ok to test

**Comment by viglesiasce** [comment]
cc @krancour for review

@c-knowles this needs a version bump of the chart

File: stable/traefik/Chart.yaml
Comment (krancour) [inline stable/traefik/Chart.yaml]: The version scheme this particular chart has observed, to date, is that we are in lock step with the semver version number of Traefik itself, and we append `-a`, `-b`, etc. to denote versions of the _chart_ for that given Traefik release. So I'd rather see this change to `1.1.2-b` since the chart continues to use Traefik 1.1.2.

File: stable/traefik/README.md
Comment (krancour) [inline stable/traefik/README.md]: To allow for the possibility of other auth mechanisms being configurable in the future, I might suggest `dasboard.auth.basic` instead.

Comment (krancour) [reply stable/traefik/README.md]: Can the description make it more clear that this is a map?

Comment (krancour) [reply stable/traefik/README.md]: Sorry... also... is there some way we can avoid possibly misleading anyone into thinking "no auth" is actually a default value? Can we use language like "unset by default; this means basic auth is disabled." No one will mistake that as the actual default value of the property.

Can you also add a section for this to the default values.yaml and leave it commented. It's easier for people to get this bit of config right if it's already present and commented; meaning they don't have to start from scratch and possibly get an indent wrong or something.

**Review by krancour — CHANGES_REQUESTED** [review]
Looks good, but want to suggest a few improvements.

File: stable/traefik/Chart.yaml
Comment (cknowles) [reply stable/traefik/Chart.yaml]: Sure ok, I was wondering about that. I'll add a note next to the version to make it more obvious.

File: stable/traefik/README.md
Comment (cknowles) [reply stable/traefik/README.md]: Yep, agree on all points. I had stated it was a map originally before I moved to the auth section but worth keeping it more explicit.

**Comment by cknowles** [comment]
@krancour fixed up those.

One thing I'm not sure of is if there is a better way to avoid template errors in the case the `auth` property is not specified. It seems helm does not do a deep merge of user and default values so if they specify `dashboard` without `auth` it will fail to render without the additional `{{- if .Values.dashboard.auth }}` (can't use `and` as it evaluates all args).

**Comment by krancour** [comment]
> One thing I'm not sure of is if there is a better way to avoid template errors in the case the auth property is not specified.

I played with this a bit... If `auth` is present but blank, it is considered a string, which is a type that won't have a field `basic`. If `auth` is absent, it is considered a nil, which also doesn't have a field `basic`. Short-circuiting doesn't seem to work with the `and` template function, because it would seem the templating engine evaluates values _before_ passing them to the function and not after...

So, awkward as it looks, @c-knowles, I think that what you've done is the best that can be managed here.

This all LGTM and feature-wise, I think this is a fantastic addition to the chart!


cc @viglesiasce @lachie83. Can we merge?

**Comment by krancour** [comment]
Err... hold off on merge, actually. I am running into some trouble with manual tests. I want to figure out what's up with that first.

**Comment by krancour** [comment]
Actually, hold off on merge until we discuss something...

As a result of this PR, anyone using this feature could have encoded, but unencrypted passwords stored in a configmap. That's frowned upon.

**Comment by krancour** [comment]
Following up on the above, I considered whether it was practical to put the entirety of `traefik.toml` in a secret instead of a configmap... it's not. The reason is that values in secrets _must_ be base64 encoded, and I can find no way to base64 the entirety of `traefik.toml`, since that itself is constructed from a complex template with lots of conditional and iterative logic.

This being said, the only practical difference between a configmap and a secret is that k8s enforces that values in a secret must be base64 encoded. i.e. The values of secrets are _not_ encrypted and aren't secure-- they are only obfuscated. @c-knowles PR already implicitly requires obfuscation of basic auth passwords because Traefik doesn't accept clear text for passwords; it accepts only values encoded with MD5, SHA1 and BCrypt.

This being the case, what's PR'ed here isn't inherently less secure than any other alternative.

So this still LGTM and as long as @viglesiasce and @lachie83 accept the above, I'd still urge us to consider merging.

**Review by krancour — APPROVED** [review]

**Comment by cknowles** [comment]
My point of view mirrors that of @krancour, the passwords are already encoded although it's likely they are still better placed in Secret if possible. I did think about this part but forgot to add a note for it. I couldn't spot a way to give Traefik multiple configurations to attempt to separate secrets from other config.

If we wanted to move entirely to Secret I think that could be achievable. The documentation is sparse but as Helm uses [sprig](https://github.com/Masterminds/sprig) so if we separated the traefik.toml template to its own file, we should be able to use the `b64enc` function a bit like `{{ include (print "/templates/traefik.toml") . | b64enc }}`.

**Comment by krancour** [comment]
We could try that... but is it wise? It will unnecessarily obfuscate _everything_ in the config.

**Comment by cknowles** [comment]
Probably not. If we agree to merge, we can file an issue with Traefik to consider adding support for separate secrets? I found this [issue about env config](https://github.com/containous/traefik/issues/7) indicating they now use [Staert](https://github.com/containous/staert) so multiple config sources should in theory be easy...

**Comment by krancour** [comment]
@c-knowles, I'll open such an issue and will consider PR'ing it myself, time-permitting. Or feel free to beat me to it. 😉

**Comment by krancour** [comment]
Opened issue https://github.com/containous/traefik/issues/1055

fwiw, imo, this PR shouldn't be dependent on getting that resolved. Once that's solved on the Traefik end, a future iteration of this chart could pretty seamlessly transition to the new/improved strategy.

**Comment by krancour** [comment]
@viglesiasce want to weigh in on this? I'd like to consider merging.

**Comment by lachie83** [comment]
@c-knowles @krancour Is there any other way to set the password at runtime rather than in the config file. It would be nice to break this out into a secret then pull it in via an ENVVAR. Eg. https://github.com/kubernetes/charts/blob/master/stable/jenkins/templates/jenkins-master-deployment.yaml#L52 If that's not possible then I'm okay with this as I don't see any other way currently.

**Comment by cknowles** [comment]
@krancour @lachie83 I found [these docs](https://docs.traefik.io/toml/#file-backend) so it seems we should be able to do something like the below in the main `traefik.toml` and add an `auth.toml` as a Secret.
```
[file]
filename = "auth.toml"
```

**Comment by krancour** [comment]
@c-knowles that looks like an interesting option that might work. Very curious to see how it pans out.

**Comment by lachie83** [comment]
@c-knowles Great. It would be worth trying

**Comment by cknowles** [comment]
Tried this out but found a [blocker on the Traefik side](https://github.com/containous/traefik/blob/master/server.go#L219), it seems it's only possible to put backend/frontend config in the separate files and not auth information.

**Comment by krancour** [comment]
> it seems it's only possible to put backend/frontend config in the separate files and not auth information

D'oh!

Ok... I think we we should proceed with this as is. As we've discussed, encoded values in a configmap aren't any less secure than the encoded (_not_ encrypted!) values that normally appear in a secret. Secrets are a more conventional place to put such information, but we can transition to doing that in the future _after_ https://github.com/containous/traefik/issues/1055 is addressed.

@lachie83 you ok with this as is?

@c-knowles, if @lachie83 is happy, then just bump the version string to `1.1.2-e`.

**Comment by lachie83** [comment]
lgtm

**Comment by lachie83** [comment]
Merge #431 first

**Comment by krancour** [comment]
Rock on! Love this feature. Thanks for all the efforts @c-knowles

**Pull request merged**
